### PR TITLE
Migrate to new Elasticsearch Java client for Spring Boot 3

### LINKED ISSUE #19934: Migrate to new Elasticsearch Java client for Spring Boot 3
##### **Overview of the issue**

Spring Boot 3.0.0 M4 introduces a [new Elasticsearch Java client](https://github.com/spring-projects/spring-boot/wiki/Spring-Boot-3.0.0-M4-Release-Notes#elasticsearch-java-client).

If you create a new project with [Spring Boot 3 support for JHipster](https://github.com/jhipster/generator-jhipster/pull/19791), it fails to compile. I'm creating this issue separately to add a bug bounty and get some help.

##### **Motivation for or Use Case**

It's necessary to upgrade our Elasticsearch support for Spring Boot 3.

##### **Reproduce the error**

```
take 21-points
jhipster jdl 21-points.jh
npm run app:start
```

Results in:

```
/Users/mraible/Downloads/21-points/src/main/java/com/okta/developer/repository/search/PointsSearchRepository.java:3: error: package org.elasticsearch.index.query does not exist
import static org.elasticsearch.index.query.QueryBuilders.queryStringQuery;
                                           ^
/Users/mraible/Downloads/21-points/src/main/java/com/okta/developer/repository/search/PointsSearchRepository.java:3: error: static import only from classes and interfaces
import static org.elasticsearch.index.query.QueryBuilders.queryStringQuery;
^
/Users/mraible/Downloads/21-points/src/main/java/com/okta/developer/repository/search/PointsSearchRepository.java:11: error: package org.elasticsearch.search.sort does not exist
import org.elasticsearch.search.sort.SortBuilder;
                                    ^
/Users/mraible/Downloads/21-points/src/main/java/com/okta/developer/repository/search/PointsSearchRepository.java:16: error: cannot find symbol
import org.springframework.data.elasticsearch.core.ElasticsearchRestTemplate;
                                                  ^
  symbol:   class ElasticsearchRestTemplate
  location: package org.springframework.data.elasticsearch.core
/Users/mraible/Downloads/21-points/src/main/java/com/okta/developer/repository/search/PointsSearchRepository.java:19: error: cannot find symbol
import org.springframework.data.elasticsearch.core.query.NativeSearchQuery;
                                                        ^
  symbol:   class NativeSearchQuery
  location: package org.springframework.data.elasticsearch.core.query
/Users/mraible/Downloads/21-points/src/main/java/com/okta/developer/repository/search/PointsSearchRepository.java:20: error: cannot find symbol
import org.springframework.data.elasticsearch.core.query.NativeSearchQueryBuilder;
                                                        ^
  symbol:   class NativeSearchQueryBuilder
  location: package org.springframework.data.elasticsearch.core.query
/Users/mraible/Downloads/21-points/src/main/java/com/okta/developer/repository/search/PointsSearchRepository.java:43: error: cannot find symbol
    private final ElasticsearchRestTemplate elasticsearchTemplate;
                  ^
  symbol:   class ElasticsearchRestTemplate
  location: class PointsSearchRepositoryInternalImpl
/Users/mraible/Downloads/21-points/src/main/java/com/okta/developer/repository/search/PointsSearchRepository.java:46: error: cannot find symbol
    PointsSearchRepositoryInternalImpl(ElasticsearchRestTemplate elasticsearchTemplate, PointsRepository repository) {
                                       ^
  symbol:   class ElasticsearchRestTemplate
  location: class PointsSearchRepositoryInternalImpl
/Users/mraible/Downloads/21-points/src/main/java/com/okta/developer/repository/search/PreferencesSearchRepository.java:3: error: package org.elasticsearch.index.query does not exist
import static org.elasticsearch.index.query.QueryBuilders.queryStringQuery;
                                           ^
/Users/mraible/Downloads/21-points/src/main/java/com/okta/developer/repository/search/PreferencesSearchRepository.java:3: error: static import only from classes and interfaces
import static org.elasticsearch.index.query.QueryBuilders.queryStringQuery;
^
/Users/mraible/Downloads/21-points/src/main/java/com/okta/developer/repository/search/PreferencesSearchRepository.java:8: error: cannot find symbol
import org.springframework.data.elasticsearch.core.ElasticsearchRestTemplate;
                                                  ^
  symbol:   class ElasticsearchRestTemplate
  location: package org.springframework.data.elasticsearch.core
/Users/mraible/Downloads/21-points/src/main/java/com/okta/developer/repository/search/PreferencesSearchRepository.java:10: error: cannot find symbol
import org.springframework.data.elasticsearch.core.query.NativeSearchQuery;
                                                        ^
  symbol:   class NativeSearchQuery
  location: package org.springframework.data.elasticsearch.core.query
/Users/mraible/Downloads/21-points/src/main/java/com/okta/developer/repository/search/PreferencesSearchRepository.java:33: error: cannot find symbol
    private final ElasticsearchRestTemplate elasticsearchTemplate;
                  ^
  symbol:   class ElasticsearchRestTemplate
  location: class PreferencesSearchRepositoryInternalImpl
/Users/mraible/Downloads/21-points/src/main/java/com/okta/developer/repository/search/PreferencesSearchRepository.java:36: error: cannot find symbol
    PreferencesSearchRepositoryInternalImpl(ElasticsearchRestTemplate elasticsearchTemplate, PreferencesRepository repository) {
                                            ^
  symbol:   class ElasticsearchRestTemplate
  location: class PreferencesSearchRepositoryInternalImpl
/Users/mraible/Downloads/21-points/src/main/java/com/okta/developer/repository/search/WeightSearchRepository.java:3: error: package org.elasticsearch.index.query does not exist
import static org.elasticsearch.index.query.QueryBuilders.queryStringQuery;
                                           ^
/Users/mraible/Downloads/21-points/src/main/java/com/okta/developer/repository/search/WeightSearchRepository.java:3: error: static import only from classes and interfaces
import static org.elasticsearch.index.query.QueryBuilders.queryStringQuery;
^
/Users/mraible/Downloads/21-points/src/main/java/com/okta/developer/repository/search/WeightSearchRepository.java:11: error: package org.elasticsearch.search.sort does not exist
import org.elasticsearch.search.sort.SortBuilder;
                                    ^
/Users/mraible/Downloads/21-points/src/main/java/com/okta/developer/repository/search/WeightSearchRepository.java:16: error: cannot find symbol
import org.springframework.data.elasticsearch.core.ElasticsearchRestTemplate;
                                                  ^
  symbol:   class ElasticsearchRestTemplate
  location: package org.springframework.data.elasticsearch.core
/Users/mraible/Downloads/21-points/src/main/java/com/okta/developer/repository/search/WeightSearchRepository.java:19: error: cannot find symbol
import org.springframework.data.elasticsearch.core.query.NativeSearchQuery;
                                                        ^
  symbol:   class NativeSearchQuery
  location: package org.springframework.data.elasticsearch.core.query
/Users/mraible/Downloads/21-points/src/main/java/com/okta/developer/repository/search/WeightSearchRepository.java:20: error: cannot find symbol
import org.springframework.data.elasticsearch.core.query.NativeSearchQueryBuilder;
                                                        ^
  symbol:   class NativeSearchQueryBuilder
  location: package org.springframework.data.elasticsearch.core.query
/Users/mraible/Downloads/21-points/src/main/java/com/okta/developer/repository/search/WeightSearchRepository.java:43: error: cannot find symbol
    private final ElasticsearchRestTemplate elasticsearchTemplate;
                  ^
  symbol:   class ElasticsearchRestTemplate
  location: class WeightSearchRepositoryInternalImpl
/Users/mraible/Downloads/21-points/src/main/java/com/okta/developer/repository/search/WeightSearchRepository.java:46: error: cannot find symbol
    WeightSearchRepositoryInternalImpl(ElasticsearchRestTemplate elasticsearchTemplate, WeightRepository repository) {
                                       ^
  symbol:   class ElasticsearchRestTemplate
  location: class WeightSearchRepositoryInternalImpl
/Users/mraible/Downloads/21-points/src/main/java/com/okta/developer/repository/search/BloodPressureSearchRepository.java:3: error: package org.elasticsearch.index.query does not exist
import static org.elasticsearch.index.query.QueryBuilders.queryStringQuery;
                                           ^
/Users/mraible/Downloads/21-points/src/main/java/com/okta/developer/repository/search/BloodPressureSearchRepository.java:3: error: static import only from classes and interfaces
import static org.elasticsearch.index.query.QueryBuilders.queryStringQuery;
^
/Users/mraible/Downloads/21-points/src/main/java/com/okta/developer/repository/search/BloodPressureSearchRepository.java:11: error: package org.elasticsearch.search.sort does not exist
import org.elasticsearch.search.sort.SortBuilder;
                                    ^
/Users/mraible/Downloads/21-points/src/main/java/com/okta/developer/repository/search/BloodPressureSearchRepository.java:16: error: cannot find symbol
import org.springframework.data.elasticsearch.core.ElasticsearchRestTemplate;
                                                  ^
  symbol:   class ElasticsearchRestTemplate
  location: package org.springframework.data.elasticsearch.core
/Users/mraible/Downloads/21-points/src/main/java/com/okta/developer/repository/search/BloodPressureSearchRepository.java:19: error: cannot find symbol
import org.springframework.data.elasticsearch.core.query.NativeSearchQuery;
                                                        ^
  symbol:   class NativeSearchQuery
  location: package org.springframework.data.elasticsearch.core.query
/Users/mraible/Downloads/21-points/src/main/java/com/okta/developer/repository/search/BloodPressureSearchRepository.java:20: error: cannot find symbol
import org.springframework.data.elasticsearch.core.query.NativeSearchQueryBuilder;
                                                        ^
  symbol:   class NativeSearchQueryBuilder
  location: package org.springframework.data.elasticsearch.core.query
/Users/mraible/Downloads/21-points/src/main/java/com/okta/developer/repository/search/BloodPressureSearchRepository.java:44: error: cannot find symbol
    private final ElasticsearchRestTemplate elasticsearchTemplate;
                  ^
  symbol:   class ElasticsearchRestTemplate
  location: class BloodPressureSearchRepositoryInternalImpl
/Users/mraible/Downloads/21-points/src/main/java/com/okta/developer/repository/search/BloodPressureSearchRepository.java:47: error: cannot find symbol
    BloodPressureSearchRepositoryInternalImpl(ElasticsearchRestTemplate elasticsearchTemplate, BloodPressureRepository repository) {
                                              ^
  symbol:   class ElasticsearchRestTemplate
  location: class BloodPressureSearchRepositoryInternalImpl
/Users/mraible/Downloads/21-points/src/main/java/com/okta/developer/repository/search/UserSearchRepository.java:3: error: package org.elasticsearch.index.query does not exist
import static org.elasticsearch.index.query.QueryBuilders.queryStringQuery;
                                           ^
/Users/mraible/Downloads/21-points/src/main/java/com/okta/developer/repository/search/UserSearchRepository.java:3: error: static import only from classes and interfaces
import static org.elasticsearch.index.query.QueryBuilders.queryStringQuery;
^
/Users/mraible/Downloads/21-points/src/main/java/com/okta/developer/repository/search/UserSearchRepository.java:7: error: cannot find symbol
import org.springframework.data.elasticsearch.core.ElasticsearchRestTemplate;
                                                  ^
  symbol:   class ElasticsearchRestTemplate
  location: package org.springframework.data.elasticsearch.core
/Users/mraible/Downloads/21-points/src/main/java/com/okta/developer/repository/search/UserSearchRepository.java:9: error: cannot find symbol
import org.springframework.data.elasticsearch.core.query.NativeSearchQuery;
                                                        ^
  symbol:   class NativeSearchQuery
  location: package org.springframework.data.elasticsearch.core.query
/Users/mraible/Downloads/21-points/src/main/java/com/okta/developer/repository/search/UserSearchRepository.java:23: error: cannot find symbol
    private final ElasticsearchRestTemplate elasticsearchTemplate;
                  ^
  symbol:   class ElasticsearchRestTemplate
  location: class UserSearchRepositoryInternalImpl
/Users/mraible/Downloads/21-points/src/main/java/com/okta/developer/repository/search/UserSearchRepository.java:25: error: cannot find symbol
    UserSearchRepositoryInternalImpl(ElasticsearchRestTemplate elasticsearchTemplate) {
                                     ^
  symbol:   class ElasticsearchRestTemplate
  location: class UserSearchRepositoryInternalImpl
/Users/mraible/Downloads/21-points/src/main/java/com/okta/developer/web/rest/PreferencesResource.java:3: error: package org.elasticsearch.index.query does not exist
import static org.elasticsearch.index.query.QueryBuilders.*;
                                           ^
/Users/mraible/Downloads/21-points/src/main/java/com/okta/developer/web/rest/BloodPressureResource.java:3: error: package org.elasticsearch.index.query does not exist
import static org.elasticsearch.index.query.QueryBuilders.*;
                                           ^
/Users/mraible/Downloads/21-points/src/main/java/com/okta/developer/web/rest/PointsResource.java:3: error: package org.elasticsearch.index.query does not exist
import static org.elasticsearch.index.query.QueryBuilders.*;
                                           ^
/Users/mraible/Downloads/21-points/src/main/java/com/okta/developer/web/rest/WeightResource.java:3: error: package org.elasticsearch.index.query does not exist
import static org.elasticsearch.index.query.QueryBuilders.*;
                                           ^
/Users/mraible/Downloads/21-points/src/main/java/com/okta/developer/web/rest/PublicUserResource.java:3: error: package org.elasticsearch.index.query does not exist
import static org.elasticsearch.index.query.QueryBuilders.*;
```

##### **Related issues**

#19791

### PR DESCRIPTION
**Pull request by frascu**
Fixes #19934

---

Please make sure the below checklist is followed for Pull Requests.

- [x] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
- [x] Tests are added where necessary
- [x] The JDL part is updated if necessary
- [x] [jhipster-online](https://github.com/jhipster/jhipster-online) is updated if necessary
- [x] Documentation is added/updated where necessary
- [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

When you are still working on the PR, consider converting it to Draft (below reviewers) and adding `skip-ci` label, you can still see CI build result at your branch.

### CODE REVIEW

File: generators/server/templates/sql/reactive/src/main/java/package/repository/EntityManager.java.ejs [hunk 1]
```diff
@@ -157,7 +157,7 @@ public class EntityManager {
      * @param entityType the entity type which holds the table name.
      * @return the number of deleted entity
      */
-    public Mono<Integer> deleteAll(Class<?> entityType) {
+    public Mono<Long> deleteAll(Class<?> entityType) {
```
Comment (mraible) [inline generators/server/templates/sql/reactive/src/main/java/package/repository/EntityManager.java.ejs:160]: Why is this change necessary? It seems unrelated to Elasticsearch. I'm fine with it, but not sure if it'll mess up database logic.

**Review by mraible — APPROVED** [review]

File: generators/server/templates/sql/reactive/src/main/java/package/repository/EntityManager.java.ejs (hunk 1 above)
Comment (frascu) [reply generators/server/templates/sql/reactive/src/main/java/package/repository/EntityManager.java.ejs:160]: In order to fix the reactive elasticsearch client I generated a reactive project.
Together with the replacement of the right package for `ReactiveElasticsearchTemplate`, I changed also the methods in this class because in several case (like this one) the latest version (`3.0.0-M6`) of `spring-data-r2dbc`  return `Mono<Integer>` instead of `Mono<Long>`.

I think this changed is necessary for the migration to Spring Boot 3, but as you wrote it is unrelated to the elastichseach.
Therefore if you want I can undo just the changes in the `EntitManager` class.

Comment (mraible) [reply generators/server/templates/sql/reactive/src/main/java/package/repository/EntityManager.java.ejs:160]: I think it's OK to make this change.

**Comment by mraible** [comment]
@frascu Can you please update your branch from `spring-boot-3.0-m4`?

**Comment by frascu** [comment]
Updated

**Comment by frascu** [comment]
@mraible @DanielFran since the failed pipelines are not related to Elasticsearch migration, can this pull request be merged?

**Comment by mraible** [comment]
I'm OK with merging it. Is that OK with you, @mshima?

**Comment by mshima** [comment]
I prefer to merge after spring-boot rc1 release which is due in 3 days 20/10.
RC1 should bring the Problem feature, and maybe we can make backend tests to pass before merging this PR.
Smaller PRs is much easier to review, the generated diff is quite hand.

@mraible unless if you plan to do some development that requires this PR.
So feel free to merge this PR and https://github.com/jhipster/generator-jhipster/pull/20011 too if you want.

**Comment by mshima** [comment]
Closing/reopen to create a new merge point

**Comment by mshima** [comment]
Is this missing reactive implementation?
Or reactive kept the same?

**Comment by mraible** [comment]
I'm merging this PR so @frascu can collect a bug bounty for his work here.

**Comment by frascu** [comment]
Bounty claimed https://opencollective.com/generator-jhipster/expenses/102550 🙂

**Comment by frascu** [comment]
@DanielFran can you approve?

**Comment by DanielFran** [comment]
@frascu approved
